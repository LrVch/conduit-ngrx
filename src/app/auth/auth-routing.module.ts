import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { AuthComponent } from './auth.component';
import { NoAuthGuard } from './no-auth-guard.service';

const routes: Routes = [
  {
    path: 'login',
    component: AuthComponent,
    canActivate: [NoAuthGuard],
    data: { title: 'conduit.menu.login', animation: 'AuthComponent' }
  },
  {
    path: 'register',
    component: AuthComponent,
    canActivate: [NoAuthGuard],
    data: { title: 'conduit.menu.register', animation: 'AuthComponent' }
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class AuthRoutingModule {}
