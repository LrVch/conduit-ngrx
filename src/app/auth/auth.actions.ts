import { Action } from '@ngrx/store';
import { Credentials } from '@app/core/models/credentials.model';
import { User, Errors } from '@app/core';

export enum AuthActionTypes {
  LoginPageAttemptLogin = '[Login Page] Attempt Login',
  LoginSuccess = '[Auth API] Login success',
  LoginFail = '[Auth API] Login fail',
  LoginAttemptCopmlete = '[Auth API] Login Attempt Copmlete',
  LoginPageSetAuthErrors = '[Auth API] Set Auth Errors',
  LoginPageClearAuthErrors = '[Login Page] Clear Auth Errors',
  LoggedLocalStorageRequest = '[LocalStorage] Login success Request',
  LoggedLocalStorage = '[LocalStorage] Login Success',
  ClearAuthErrors = '[Login Page] Clear Auth Errors',
  SettingsPageLogoutAction = '[Settings Page] Logout',
  LogoutConfirm = '[Auth] Logout Confirm',
  LogoutAction = '[Auth] Logout',
  AuthAttemptToGetUser = '[Auth] Try To Get User From Api',
  SetReturnUrl = '[Auth] Set Return Url',
  ClearReturnStateFromRouteChange = '[Auth] Clear Return State From Route Change',

  UpdateUserRequest = '[Auth API] Update User Request',
  UpdateUserSuccess = '[Auth API] Update User Success',
  UpdateUserFail = '[Auth API] Update User Fail',
  SetUpdateUserErrors = '[Auth API] Set Update User Errors'
}

export class SetReturnUrl implements Action {
  readonly type = AuthActionTypes.SetReturnUrl;

  constructor(public payload: { returnUrl: string }) { }
}

export class ClearReturnStateFromRouteChange implements Action {
  readonly type = AuthActionTypes.ClearReturnStateFromRouteChange;
}

export class AuthAttemptToGetUser implements Action {
  readonly type = AuthActionTypes.AuthAttemptToGetUser;
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.LoginSuccess;

  constructor(public payload: { user: User }) { }
}

export class LoggedLocalStorage implements Action {
  readonly type = AuthActionTypes.LoggedLocalStorage;

  constructor(public payload: { user: User }) { }
}

export class LoggedLocalStorageRequest implements Action {
  readonly type = AuthActionTypes.LoggedLocalStorageRequest;
}

export class LoginFail implements Action {
  readonly type = AuthActionTypes.LoginFail;

  constructor(public payload: { authErrors: Errors }) { }
}

export class LoginAttemptCopmlete implements Action {
  readonly type = AuthActionTypes.LoginAttemptCopmlete;
}

export class SettingsPageLogoutAction implements Action {
  readonly type = AuthActionTypes.SettingsPageLogoutAction;

  constructor(public payload: { question: string }) { }
}

export class LogoutConfirm implements Action {
  readonly type = AuthActionTypes.LogoutConfirm;
}

export class LogoutAction implements Action {
  readonly type = AuthActionTypes.LogoutAction;

  constructor(public payload?: { path?: string }) { }
}

export class LoginPageAttemptLogin implements Action {
  readonly type = AuthActionTypes.LoginPageAttemptLogin;

  constructor(public payload: { authType: string, credentials: Credentials }) { }
}

export class LoginPageClearAuthErrors implements Action {
  readonly type = AuthActionTypes.LoginPageClearAuthErrors;
}

export class ClearAuthErrors implements Action {
  readonly type = AuthActionTypes.ClearAuthErrors;
}


export class LoginPageSetAuthErrors implements Action {
  readonly type = AuthActionTypes.LoginPageSetAuthErrors;

  constructor(public payload: { authErrors: Errors }) { }
}

export class SetUpdateUserErrors implements Action {
  readonly type = AuthActionTypes.SetUpdateUserErrors;

  constructor(public payload: { authErrors: Errors }) { }
}

export class UpdateUserRequest implements Action {
  readonly type = AuthActionTypes.UpdateUserRequest;

  constructor(public payload: { user: User }) { }
}

export class UpdateUserSuccess implements Action {
  readonly type = AuthActionTypes.UpdateUserSuccess;

  constructor(public payload: { user: User }) { }
}

export class UpdateUserFail implements Action {
  readonly type = AuthActionTypes.UpdateUserFail;

  constructor(public payload: { errors: Errors }) { }
}

export type AuthActions =
  SetReturnUrl
  | ClearReturnStateFromRouteChange
  | AuthAttemptToGetUser
  | LoginSuccess
  | LoginFail
  | SettingsPageLogoutAction
  | LogoutConfirm
  | LogoutAction
  | LoginAttemptCopmlete
  | LoginPageAttemptLogin
  | LoggedLocalStorage
  | LoggedLocalStorageRequest
  | LoginPageClearAuthErrors
  | ClearAuthErrors
  | LoginPageSetAuthErrors
  | UpdateUserRequest
  | UpdateUserSuccess
  | UpdateUserFail
  | SetUpdateUserErrors;


