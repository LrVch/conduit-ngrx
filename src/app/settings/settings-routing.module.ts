import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { AuthGuard } from '@app/core';
import { SettingsComponent } from './settings/settings.component';
import { CanDeactivateGuard } from '@app/core/services/can-deactivate.guard';

const routes: Routes = [
  {
    path: '',
    component: SettingsComponent,
    canActivate: [AuthGuard],
    canDeactivate: [CanDeactivateGuard]
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class SettingsRoutingModule {}
