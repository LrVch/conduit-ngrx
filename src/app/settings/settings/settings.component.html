<div class="settings-page page">
  <div class="container">
    <div class="row">
      <div class="col-md-6 offset-md-3 col-xs-12">

        <h1 class="text-xs-center">Settings</h1>

        <app-list-error [errors]="errors$ | async"></app-list-error>

        <app-settings-form *ngIf="(user$ | async)"
                           [isSubmitting]="isUpdating$ | async"
                           (updateUser)="onUpdateUser($event)"
                           [user]="user$ | async"></app-settings-form>

        <div class="d-flex justify-content-center mt-5">
          <mat-spinner *ngIf="!(user$ | async)"></mat-spinner>
        </div>

        <div *ngIf="(user$ | async)">
          <hr />
          <button (click)="logout()"
                  mat-stroked-button
                  color="primary">Click here to logout</button>
        </div>
        <br>
        <br>

      </div>
    </div>
  </div>
</div>