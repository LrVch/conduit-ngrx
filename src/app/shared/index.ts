export * from './article/article.component';
export * from './articles-list/articles-list.component';
export * from './content-tabs/content-tabs.component';
export * from './tags/tags.component';
export * from './favorite/favorite.component';
export * from './layout';
export * from './shared.module';
export * from './utils';
export * from './list-error/list-error.component';
export * from './show-authed.directive';
export * from './user-info/user-info.component';
export * from './follow-button/follow-button.component';
export * from './tag-list/tag-list.component';
export * from './comments/comments.component';
export * from './base-from/base-from.component';
