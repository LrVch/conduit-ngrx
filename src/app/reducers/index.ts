import {
  ActionReducerMap,
  MetaReducer
} from '@ngrx/store';
import { environment } from '@env/environment';
import { AuthState, authReducer } from '@app/auth/auth.reducer';
import { ArticlesState, articlesReducer } from '@app/articles/articles.reducer';
import { storeFreeze } from 'ngrx-store-freeze';
import { routerReducer, RouterReducerState } from '@ngrx/router-store';
import { ArticlesConfigState, articlesConfigReducer } from '@app/articles/articlesConfig.reducer';
import { ProfileState, profileReducer } from '@app/profile/profile.reducer';
import { layoutReducer, LayoutState } from '@app/layout/layout.reducer';
import { EditorState, editorReducer } from '@app/editor/editor.reducer';
import { ArticleState, articleReducer } from '@app/article/article.reducer';
import { AppSettingsState } from '@app/core/models/app-settings.model';
import { appSettingsReducer } from '@app/appSettings/app-settings.reducer';
import { debug } from './meta-reducers/debug.reducer';

import { initStateFromLocalStorage } from './meta-reducers/init-state-from-local-storage.reducer';


export interface AppState {
  auth: AuthState;
  articles: ArticlesState;
  router: RouterReducerState;
  articlesConfig: ArticlesConfigState;
  profile: ProfileState;
  layout: LayoutState;
  editor: EditorState;
  article: ArticleState;
  appSettings: AppSettingsState;
}

export const reducers: ActionReducerMap<AppState> = {
  auth: authReducer,
  articles: articlesReducer,
  router: routerReducer,
  articlesConfig: articlesConfigReducer,
  profile: profileReducer,
  layout: layoutReducer,
  editor: editorReducer,
  article: articleReducer,
  appSettings: appSettingsReducer
};


export const metaReducers: MetaReducer<AppState>[] = [initStateFromLocalStorage];

if (!environment.production) {
  metaReducers.unshift(storeFreeze);
}
