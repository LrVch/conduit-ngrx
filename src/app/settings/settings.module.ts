import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SettingsComponent } from './settings/settings.component';
import { SettingsRoutingModule } from './settings-routing.module';
import { SettingsFormComponent } from './settings-form/settings-form.component';
import { SharedModule } from '@app/shared';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { HttpClient } from '@angular/common/http';
import { environment } from '@env/environment.prod';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    SettingsRoutingModule,
    TranslateModule.forChild({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient]
      },
      isolate: true
    }),
  ],
  declarations: [SettingsComponent, SettingsFormComponent]
})
export class SettingsModule { }

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.i18nPrefix}/assets/i18n/examples/`,
    '.json'
  );

