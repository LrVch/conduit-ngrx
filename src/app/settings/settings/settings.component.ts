import { Component, OnInit } from '@angular/core';
import { AppState } from '@app/reducers';
import { Store, select } from '@ngrx/store';
import { SettingsPageLogoutAction, UpdateUserRequest, ClearAuthErrors } from '@app/auth/auth.actions';
import { Observable } from 'rxjs';
import { Errors, User } from '@app/core';
import { selectUser, selectAuthErrors, selectUserUpdatingInfo } from '@app/auth/auth.selectors';
import { CanComponentDeactivate } from '@app/core/services/can-deactivate.guard';
import { DialogService } from '@app/core/services/dialog.service';
import { TranslateService } from '@ngx-translate/core';
import { switchMap, map } from 'rxjs/operators';

@Component({
  selector: 'app-settings',
  templateUrl: './settings.component.html'
})
export class SettingsComponent implements OnInit, CanComponentDeactivate {
  errors$: Observable<Errors>;
  isUpdating$: Observable<boolean>;
  user$: Observable<User>;
  wasChanged = false;

  constructor(
    private store: Store<AppState>,
    private dialog: DialogService,
    private translateService: TranslateService
  ) { }

  ngOnInit() {
    this.user$ = this.store.pipe(select(selectUser));
    this.isUpdating$ = this.store.pipe(select(selectUserUpdatingInfo));
    this.errors$ = this.store.pipe(select(selectAuthErrors));

    this.store.dispatch(new ClearAuthErrors());
  }

  logout() {
    this.store.dispatch(new SettingsPageLogoutAction({ question: 'Are you sure you want to quit?' }));
  }

  onUpdateUser(user: User) {
    this.store.dispatch(new UpdateUserRequest({ user }));
  }

  onWasChanged(wasChanged: boolean) {
    this.wasChanged = wasChanged;
  }

  canDeactivate(): Observable<boolean> | boolean {
    if (!this.wasChanged) {
      return true;
    }

    return this.translateService.get('conduit.settings.goAwayWarning').pipe(
      map(question => this.dialog.confirmation({
        data: { question: question },
      })),
      switchMap(ref => ref.afterClosed()),
    );
  }
}
