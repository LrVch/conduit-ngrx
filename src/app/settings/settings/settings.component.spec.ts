import { async, ComponentFixture } from '@angular/core/testing';
import { ConfigureFn, configureTests, getUser } from '@app/lib/testing';
import { SettingsComponent } from './settings.component';
import {
  DebugElement,
  Component,
  Output,
  EventEmitter,
  Input
} from '@angular/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';
import { Store, StoreModule, combineReducers } from '@ngrx/store';
import * as AuthActions from '@app/auth/auth.actions';
import * as fromRoot from '@app/reducers';
import * as fromAuth from '@app/auth/auth.reducer';
import { hot, cold } from 'jasmine-marbles';
import { MaterialModule } from '@app/shared';
import { DialogService } from '@app/core/services/dialog.service';
import { User } from '@app/core';
import { of, Observable } from 'rxjs';
import { TranslateModule, TranslateService } from '@ngx-translate/core';

@Component({
  selector: 'app-list-error',
  template: ``
})
class TestHostListErrorComponent {
  @Input('errors') errors: any;
}

@Component({
  selector: 'app-settings-form',
  template: `
    <div id="update" (click)="onUpdateUser()"></div>
    <div id="change" (click)="onWasChanged()"></div>
  `
})
class TestHostFrotmComponent {
  @Input('isSubmitting') isSubmitting: any;
  @Input('user') user: User = getUser();
  @Output() updateUser = new EventEmitter<any>();
  @Output() wasChanged = new EventEmitter<boolean>();

  onUpdateUser() {
    this.updateUser.emit(this.user);
  }

  onWasChanged() {
    this.wasChanged.emit(true);
  }
}

describe('SettingsComponent', () => {
  let component: SettingsComponent;
  let fixture: ComponentFixture<SettingsComponent>;
  let de: DebugElement;
  let store: Store<fromAuth.AuthState>;
  let user: User;
  let dialogService: DialogService;
  let translateService: TranslateService;

  class MockDialogService {
    confirmation = jest.fn();
  }

  beforeEach(async(() => {
    user = getUser();
    const configure: ConfigureFn = testBed => {
      testBed.configureTestingModule({
        declarations: [
          SettingsComponent,
          TestHostFrotmComponent,
          TestHostListErrorComponent
        ],
        imports: [
          TranslateModule.forRoot(),
          MaterialModule,
          NoopAnimationsModule,
          StoreModule.forRoot({
            ...fromRoot.reducers,
            feature: combineReducers(fromAuth.authReducer)
          })
        ],
        providers: [
          {
            provide: DialogService,
            useClass: MockDialogService
          },
          TranslateService
        ]
      });
    };

    configureTests(configure).then(testBed => {
      fixture = testBed.createComponent(SettingsComponent);
      component = fixture.componentInstance;
      de = fixture.debugElement;
      store = testBed.get(Store);
      fixture.detectChanges();

      dialogService = testBed.get(DialogService);
      translateService = testBed.get(TranslateService);

      spyOn(store, 'dispatch').and.callThrough();
    });
  }));

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should compile', () => {
    (<any>expect(fixture)).toMatchSnapshot();
  });

  it('should compile and show user @Input', () => {
    store.dispatch(new AuthActions.LoginSuccess({ user }));
    fixture.detectChanges();

    component.user$.subscribe(res => {
      (<any>expect(fixture)).toMatchSnapshot();
    });
  });

  it('should get user', () => {
    component.ngOnInit();

    const expected = cold('a', { a: null });

    expect(component.user$).toBeObservable(expected);
  });

  it('should get isUpdating', () => {
    component.ngOnInit();

    const expected = cold('a', { a: false });

    expect(component.isUpdating$).toBeObservable(expected);
  });

  it('should get errors', () => {
    component.ngOnInit();

    const expected = cold('a', { a: { errors: {} } });

    expect(component.errors$).toBeObservable(expected);
  });

  it('should dispatch "SettingsPageLogoutAction"', () => {
    component.logout();

    expect(store.dispatch).toHaveBeenCalledWith(
      new AuthActions.SettingsPageLogoutAction({
        question: 'conduit.settings.logouQuestion'
      })
    );
  });

  it('should dispatch "UpdateUserRequest"', () => {
    store.dispatch(new AuthActions.LoginSuccess({ user }));
    expect(store.dispatch).toHaveBeenCalledWith(
      new AuthActions.LoginSuccess({ user })
    );

    fixture.detectChanges();

    component.ngOnInit();

    const updateDe = de
      .query(By.css('app-settings-form'))
      .query(By.css('#update'));

    updateDe.triggerEventHandler('click', null);

    expect(store.dispatch).toHaveBeenCalledWith(
      new AuthActions.UpdateUserRequest({ user })
    );
  });

  it('should set wasChanged to true', () => {
    store.dispatch(new AuthActions.LoginSuccess({ user }));
    expect(store.dispatch).toHaveBeenCalledWith(
      new AuthActions.LoginSuccess({ user })
    );

    fixture.detectChanges();

    const updateDe = de
      .query(By.css('app-settings-form'))
      .query(By.css('#change'));

    updateDe.triggerEventHandler('click', null);

    expect(component.wasChanged).toBeTruthy();
  });

  it("shouldn't show dialog", () => {
    component.wasChanged = false;

    expect(component.canDeactivate()).toBeTruthy();
    expect(dialogService.confirmation).not.toHaveBeenCalled();
  });

  it('should show dialog and return true', () => {
    store.dispatch(new AuthActions.LoginSuccess({ user }));
    component.wasChanged = true;

    spyOn(translateService, 'get').and.returnValue(
      of("You'll lost the data you have changed!")
    );

    dialogService.confirmation = jest.fn(() => ({
      afterClosed() {
        return of(true);
      }
    }));

    (<Observable<boolean>>component.canDeactivate()).subscribe(res => {
      expect(res).toBeTruthy();
    });
    expect(dialogService.confirmation).toHaveBeenCalledWith({
      data: { question: "You'll lost the data you have changed!" }
    });
  });

  it('should show dialog and return true', () => {
    store.dispatch(new AuthActions.LoginSuccess({ user }));
    component.wasChanged = true;

    spyOn(translateService, 'get').and.returnValue(
      of("You'll lost the data you have changed!")
    );

    dialogService.confirmation = jest.fn(() => ({
      afterClosed() {
        return of(false);
      }
    }));

    (<Observable<boolean>>component.canDeactivate()).subscribe(res => {
      expect(res).toBeFalsy();
    });
    expect(dialogService.confirmation).toHaveBeenCalledWith({
      data: { question: "You'll lost the data you have changed!" }
    });
  });
});
