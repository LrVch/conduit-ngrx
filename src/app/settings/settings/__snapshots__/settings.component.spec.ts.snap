// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsComponent should compile 1`] = `
<app-settings
  dialog={[Function MockDialogService]}
  errors$={[Function Store]}
  isUpdating$={[Function Store]}
  store={[Function Store]}
  user$={[Function Store]}
  wasChanged="false"
>
  <div
    class="settings-page page"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3 col-xs-12"
        >
          <h1
            class="text-xs-center"
          >
            Settings
          </h1>
          <app-list-error
            ng-reflect-errors="[object Object]"
          />
          
          <div
            class="d-flex justify-content-center mt-5"
          >
            
            <mat-spinner
              class="mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation _mat-animation-noopable"
              mode="indeterminate"
              role="progressbar"
              style="width: 100px; height: 100px;"
            >
              <svg
                focusable="false"
                ng-reflect-ng-switch="true"
                preserveAspectRatio="xMidYMid meet"
                style="width: 100px; height: 100px;"
                viewBox="0 0 100 100"
              >
                
                <circle
                  cx="50%"
                  cy="50%"
                  r="45"
                  style="stroke-dasharray: 282.7433388230814px; stroke-width: 10%;"
                />
                
              </svg>
            </mat-spinner>
          </div>
          
          <br />
          <br />
        </div>
      </div>
    </div>
  </div>
</app-settings>
`;

exports[`SettingsComponent should compile and show user @Input 1`] = `
<app-settings
  dialog={[Function MockDialogService]}
  errors$={[Function Store]}
  isUpdating$={[Function Store]}
  store={[Function Store]}
  user$={[Function Store]}
  wasChanged="false"
>
  <div
    class="settings-page page"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3 col-xs-12"
        >
          <h1
            class="text-xs-center"
          >
            Settings
          </h1>
          <app-list-error
            ng-reflect-errors="[object Object]"
          />
          
          <app-settings-form
            ng-reflect-is-submitting="false"
            ng-reflect-user="[object Object]"
          >
            <div
              id="update"
            />
            <div
              id="change"
            />
          </app-settings-form>
          <div
            class="d-flex justify-content-center mt-5"
          >
            
          </div>
          
          <div>
            <hr />
            <button
              class="mat-stroked-button mat-primary _mat-animation-noopable"
              color="primary"
              mat-stroked-button=""
              ng-reflect-color="primary"
            >
              <span
                class="mat-button-wrapper"
              >
                Click here to logout
              </span>
              <div
                class="mat-button-ripple mat-ripple"
                matripple=""
                ng-reflect-centered="false"
                ng-reflect-disabled="false"
                ng-reflect-trigger="[object HTMLButtonElement]"
              />
              <div
                class="mat-button-focus-overlay"
              />
            </button>
          </div>
          <br />
          <br />
        </div>
      </div>
    </div>
  </div>
</app-settings>
`;
